nan.h

```diff
@@ -1591,39 +1592,43 @@ class Callback {
 
 /* abstract */ class AsyncProgressWorker : public AsyncWorker {
  public:
+
+  typedef struct asyncdata_t {
+    char* data;
+    size_t size;
+    AsyncProgressWorker* worker;
+  } asyncdata_t;
+
+  typedef std::deque<asyncdata_t *> asyncqueue_t;
+
   explicit AsyncProgressWorker(Callback *callback_)
-      : AsyncWorker(callback_), asyncdata_(NULL), asyncsize_(0) {
-    async = new uv_async_t;
+      : AsyncWorker(callback_) {
     uv_async_init(
         uv_default_loop()
-      , async
+      , &async
       , AsyncProgress_
     );
-    async->data = this;
+    async.data = this;
 
     uv_mutex_init(&async_lock);
   }
 
   virtual ~AsyncProgressWorker() {
```

Comment:
Yes, probably. Although is not meant to be stack allocated either. The only legitimate use is to feed it heap allocated to AsyncQueWorker for subsequent deletion.

On November 10, 2015 10:15:09 PM GMT+02:00, Ben Noordhuis <notifications@github.com> wrote:
>>        , AsyncProgress_
>>      );
>> -    async->data = this;
>> +    async.data = this;
>>  
>>      uv_mutex_init(&async_lock);
>>    }
>>  
>>    virtual ~AsyncProgressWorker() {
>
>/cc @kkoopa - Hadn't realized this before but this destructor should
>probably be private, with Destroy() being the only way to delete an
>instance.  This:
>```c++
>{
>  AsyncProgressWorker temp(callback);
>}
>```
>Is unsound because the `uv_async_t` handle is not closed.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/nodejs/nan/pull/515/files#r44458627